### ISSUE TITLE
Unable to transmit "large" byte objects with twisted/python3

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#9324]
|[<img alt="pwuertz's avatar" src="https://avatars.githubusercontent.com/u/1819283?s=50" width="50" height="50">](https://github.com/pwuertz)| @pwuertz reported|
|-|-|
|Trac ID|trac#9324|
|Type|defect|
|Created|2017-11-08 15:16:30Z|

When sending a "large" blob of bytes on python3 using twisted.web the server is extremely(!) slow and consumes 100% CPU. Sending 100MB takes about 10s, sending 130MB goes up to a whopping 20s, so this might get worse exponentially. Here is some example code:

```
from twisted.internet import reactor, endpoints
from twisted.web import server, resource


class TestResource(resource.Resource):
    isLeaf = True
    SIZE_MB = 100
    TEST_DATA = ("0" * (SIZE_MB * 1024**2)).encode()

    def render(self, request):
        return TestResource.TEST_DATA


endpoints.serverFromString(reactor, "tcp:9000").listen(server.Site(TestResource()))
print("Listening on localhost:9000")
reactor.run()
```

According to cProfile the time is spent in `lazyByteSlice` from [twisted.python.compat](https://github.com/twisted/twisted/blob/trunk/src/twisted/python/compat.py). For py3 this function uses full copies instead of memory views when presenting slices. I did notice the comment concerning compatibility issues with py2, but apparently copying is not really a viable solution either.

Using `memoryview` slices instead of copying `object` does fix the problem for me.

<details><summary>Searchable metadata</summary>

```
trac-id__9324 9324
type__defect defect
reporter__pwuertz pwuertz
priority__normal normal
milestone__None None
branch__ 
branch_author__ 
status__closed closed
resolution__fixed fixed
component__core core
keywords__python3_lazyByteSlice python3 lazyByteSlice
time__1510154190947237 1510154190947237
changetime__1510217234407157 1510217234407157
version__None None
owner__Mark_Williams__mrw_____ Mark Williams <mrw@...>

```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="pwuertz's avatar" src="https://avatars.githubusercontent.com/u/1819283?s=50" width="50" height="50">](https://github.com/pwuertz)<a name="note_2"></a>|@pwuertz commented|
|-|-|

This solves the problem for me:
```
def memory_view_slice(o, offset=0, size=None):
    return memoryview(o)[offset:None if size is None else offset+size]
import twisted.python.compat
twisted.python.compat.lazyByteSlice = memory_view_slice
```

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)<a name="note_3"></a>|@itamarst commented|
|-|-|

We didn't use memoryview because of https://bugs.python.org/issue15945 apparently, but that's been fixed since Python 3.3. So it seems reasonable to switch to memoryview unconditionally.

And, maybe someone should skim the Python 3 compat layer and make sure there aren't any other timebombs we left lying around when we did the intial port :(

**Comment by twisted-trac** [comment]
|[<img alt="markrwilliams's avatar" src="https://avatars.githubusercontent.com/u/704247?s=50" width="50" height="50">](https://github.com/markrwilliams)<a name="note_4"></a>|@markrwilliams commented|
|-|-|

Replying to [itamarst](#note_3):
> We didn't use memoryview because of https://bugs.python.org/issue15945 apparently, but that's been fixed since Python 3.3. 

The strange idiom encouraged by https://bugs.python.org/issue15958 indeed eliminates the additional copy.  Consider:

```python
import sys
SIZE_MB = 768
TEST_DATA = b"A" * (SIZE_MB * 1024**2)
OFFSET = len(TEST_DATA) // 2

if sys.argv[1:2] == ['slow']:
    x = TEST_DATA[OFFSET:] + b"xyz"
else:
    x = b''.join([memoryview(TEST_DATA)[OFFSET:], b"xyz"])
```

Running this under `ltrace` without using `memoryview` shows two copies for `TEST_DATA`'s slice:

```
$ ltrace python3.6 copytest.py slow |& grep AAA
memcpy(0x7f7ac3a08030, "AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA"..., 402653184) = 0x7f7ac3a08030
memcpy(0x7f7aaba07030, "AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA"..., 402653184) = 0x7f7aaba07030
```

Whereas using `memoryview` shows a single copy:

```
$ ltrace python3.6 copytest.py fast |& grep AAA
memcpy(0x7f47d6744030, "AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA"..., 402653184) = 0x7f47d6744030
```

> So it seems reasonable to switch to memoryview unconditionally.

Unfortunately, `b''.join` does not work with `memoryview` on Python 3.3, so Twisted either needs to drop support for it or switch back to the slow way just for that version.  I prefer dropping support for 3.3.

The comment on `twisted.python.compat.lazyByteSlice` also [claims that memoryview breaks PyOpenSSL.](https://github.com/twisted/twisted/blob/501975b2240970447d2c41184198becdd7b2e413/src/twisted/python/compat.py)  I think the comment is wrong for two reasons:

1. PyOpenSSL uses cryptography now, which in turn uses cffi's `from_buffer` to slurp in binary data.  `from_buffer` fully supports the new buffer protocol used by `memoryview`.
2. `lazyByteSlice` is used in `FileDescriptor` implementations.  For PyOpenSSL to be involved, Twisted would have to use wrapped sockets; but even when this comment was written, Twisted was using Memory BIOs.

> And, maybe someone should skim the Python 3 compat layer and make sure there aren't any other timebombs we left lying around when we did the intial port :(

There's no single compat layer unfortunately.  I guess the best we can do is increase test coverage and make sure http://speed.twistedmatrix.com is useful.

**Comment by twisted-trac** [comment]
|[<img alt="markrwilliams's avatar" src="https://avatars.githubusercontent.com/u/704247?s=50" width="50" height="50">](https://github.com/markrwilliams)<a name="note_5"></a>|@markrwilliams commented|
|-|-|

https://github.com/twisted/twisted/pull/926

**Comment by twisted-trac** [comment]
|[<img alt="markrwilliams's avatar" src="https://avatars.githubusercontent.com/u/704247?s=50" width="50" height="50">](https://github.com/markrwilliams)<a name="note_7"></a>|@markrwilliams set owner to @markrwilliams<br>@markrwilliams set status to `closed`|
|-|-|

In changeset 39554dd43ab57b352110cf66344bd77158c55cea
```
#!CommitTicketReference repository="" revision="39554dd43ab57b352110cf66344bd77158c55cea"
Merge pull request #926 from twisted/9324-memoryview-buffer

Author: markrwilliams
Reviewer: moshez
Fixes: ticket:9324

Fix speed regression in Python 3.4+ by switching write buffers to use memoryview.
```
